classifier_and_summarizer/summarizer/text_media_matcher.py

```diff
@@ -0,0 +1,173 @@
+            self.media_preference_for_sentence[i].sort(
+                key=lambda x: self.similarity_matrix[i][x], reverse=True)
+            self.sentence_preference_for_media[i].sort(
+                key=lambda x: self.similarity_matrix[x][i], reverse=True)
+
+    def _gale_shapley_matching(self):
+        '''
+        Finds the stable matching between the text and media
+        Given two (n,n) matrices of preferences for the set of Sentences, Media
+        finds the stable matching by running the gale-shapley
+        matching algorithm
+        Returns : a list of tuples where each tuple (x,y) means
+        x = index of sentence
+        y = index of media
+        Thus, it returns the indices matched as a list of tuples
+        '''
+        n = self.set_size
+
+        # Make the matching optimal for the Media
+        some_media_is_unmatched = True
+
+        # -1 denotes it is currently unmatched
+        sentence_matched_for_media = [-1 for i in range(n)]
+        media_matched_for_sentence = [-1 for i in range(n)]
+
+        def media_rank_in_sentence_preference(sentence_index, media_index):
+            for i in range(n):
+                if (self.media_preference_for_sentence[sentence_index][i]
+                        == media_index):
+                    return i
+
+        def sentence_has_better_preference(sentence_index, unmatched_index):
+            return media_rank_in_sentence_preference(
+                sentence_index, unmatched_index) < \
+                    media_rank_in_sentence_preference(
+                sentence_index, media_matched_for_sentence[i])
+
+        while some_media_is_unmatched:
+            unmatched_media_index = -1  # no index found currently
+            for i in range(n):
+                if sentence_matched_for_media[i] == -1:
+                    unmatched_media_index = i
+                    break
+
+            for i in self.sentence_preference_for_media[unmatched_media_index]:
+                # the sentence is unmatched
+                if media_matched_for_sentence[i] == -1:
+                    # we can match the sentence directly
+                    media_matched_for_sentence[i] = unmatched_media_index
+                    sentence_matched_for_media[unmatched_media_index] = i
+                    break
+                elif sentence_has_better_preference(i, unmatched_media_index):
+                    # i prefers the current media better
+                    # unmatch media currently matched for sentence i
+                    sentence_matched_for_media[media_matched_for_sentence[i]] \
+                        = -1
+                    sentence_matched_for_media[unmatched_media_index] = i
+                    media_matched_for_sentence[i] = unmatched_media_index
+                    break
+
+            some_media_is_unmatched = False
+            for i in range(n):
+                if sentence_matched_for_media[i] == -1:
+                    some_media_is_unmatched = True
```

Comment:
We could keep a count of the unmatched media and decrement it when a new media gets matched (eg in the if condition of the for loop above). That might remove the need for this additional for-loop. WDYT?

Reply:
Updated accordingly - thanks for the suggestion!